core/src/main/java/org/infinispan/util/concurrent/locks/OwnableReentrantLock.java

```diff
@@ -40,6 +42,7 @@ protected final Object currentRequestor() {
    }
 
    private void setCurrentRequestor(Object requestor) {
+      assert requestorOnStack.get() == null;
```

Comment:
I see some other potential optimizations here:
* Only set/unset the TL in `lock(Object)` if the CAS operation fails.
* `unlock(Object)` shouldn't need the TL at all.

And a possible problem: `lockInterruptibly()` `tryLock()` should throw an `UnsupportedOperationException`, because they can't set the owner.